infra/src/main/kotlin/fi/fta/geoviite/infra/linking/switches/SwitchLinkingService.kt

```diff
@@ -1337,198 +1368,214 @@ fun linkJointToEdge(
-                            validJointsOnEdges.none { validJointOnEdge ->
-                                validJointOnEdge.locationTrackId == locationTrackId
-                            }
-                        isUnhandledLocationTrack
-                    }
-                    .map { (locationTrack, jointsOnLocationTrack) ->
-                        val jointNumbersOnLocationTrack =
-                            jointsOnLocationTrack
-                                .map { jointOnLocationTrack -> jointOnLocationTrack.jointNumber }
-                                .distinct()
-                        val missingJoints = structureJointSequence.minus(jointNumbersOnLocationTrack)
-                        locationTrack to missingJoints
+fun findMissingJoints(jointSequence: List<JointNumber>, jointsOnLocationTrack: List<JointOnEdge>): List<JointNumber> {
+    val jointNumbersOnLocationTrack =
+        jointsOnLocationTrack.map { jointOnLocationTrack -> jointOnLocationTrack.jointNumber }.distinct()
+    val missingJoints = jointSequence.minus(jointNumbersOnLocationTrack)
+    val missingSomeJoint = missingJoints.size > 0
+    val missingAllJoints = missingJoints.size == jointSequence.size
+    return if (missingSomeJoint && !missingAllJoints) missingJoints else emptyList()
+}
+
+/**
+ * Tries to create missing joints.
+ *
+ * @return completed joint sequence if it is possible, otherwise empty list
+ */
+fun completeJointSequence(
+    fittedSwitch: FittedSwitch,
+    jointSequence: List<JointNumber>,
+    locationTrackId: IntId<LocationTrack>,
+    jointsOnLocationTrack: List<JointOnEdge>,
+): List<JointOnEdge> {
+    val middleJointNumbers = jointSequence.drop(1).dropLast(1)
+    val missingJointNumbers = findMissingJoints(jointSequence, jointsOnLocationTrack)
+    val middleJointIsMissing =
+        missingJointNumbers.any { missingJointNumber -> middleJointNumbers.contains(missingJointNumber) }
+    val middleJointOnEdge =
+        jointsOnLocationTrack.firstOrNull { jointOnEdge -> middleJointNumbers.contains(jointOnEdge.jointNumber) }
+
+    val completedJointsOnLocationTrack =
+        if (middleJointIsMissing || middleJointOnEdge == null) {
+            // Middle joint is missing and it cannot be created automatically
+            listOf<JointOnEdge>()
+        } else {
+            // Try to create missing joints
+            val newJoints =
+                missingJointNumbers
+                    .map { missingJointNumber ->
+                        val useReverseOrder = middleJointOnEdge.direction == RelativeDirection.Against
+                        val sortedJointSequence = if (useReverseOrder) jointSequence.reversed() else jointSequence
+                        val missingJointIsBeforeMiddleJoint =
+                            sortedJointSequence.indexOf(missingJointNumber) <
+                                sortedJointSequence.indexOf(middleJointOnEdge.jointNumber)
+                        val newJointLocationM =
```

Comment:
Kummista m-arvoista tässä pitäisi puhua, edgen vai raiteen? Jos edgeltä pyytää m-arvoja ne ei ole koko raiteen m-arvoja vaan vain edgen sisäisiä. Tuo JointOnEdge sanoo vain `m`, mikä jättää myös vähän auki että kummista m-arvoista on puhe. Luokkanimi toki puhuu edgestä mutta samainen luokka sisältää myös koko raiteen geometrian.

Reply:
Also, note to self: noiden m-arvo-muuttujien olisi ehkä hyvä kertoa nimessään että minkä scopen m on kyseessä. Vaikka konteksti (pyytääkö m-arvoa segmentiltä, edgeltä vai raiteelta) tavallaan sen kertookin, niin noissa on liian helppo mennä sekaisin.